higher/index.ts

```diff
@@ -0,0 +1,2314 @@
+   *
+   * @param {string} contactName contact name
+   * @param {Object[]} contactGroups contact's linked group information
+   *
+   * @private
+   */
+  async #parseContactInfo(contactName: string, contactGroups: groupObjType[]) {
+    let linkedName = this.getLinkedName();
+    let linkedIdkeys = this.#resolveIDs([Higher.LINKED]);
+    //console.log("PARSING CONTACT");
+  
+    // check if "linked" backpointer will be replaced with "contact" backpointer
+    let contactLevelIDs = [];
+    for (let contactGroup of contactGroups) {
+      let deepCopy = JSON.parse(JSON.stringify(contactGroup));
+      if (this.#groupContains(deepCopy, Higher.LINKED)) {
+        contactLevelIDs.push(deepCopy.id);
+      }
+    }
+  
+    for (let contactGroup of contactGroups) {
+      let updatedContactGroup = this.#groupReplace(contactGroup, Higher.LINKED, Higher.CONTACTS);
+      // "linked" backpointer was replaced with "contact" backpointer
+      // set contactLevel field = true
+      if (contactLevelIDs.includes(updatedContactGroup.id)) {
+        updatedContactGroup.value.contactLevel = true;
+      }
+      // create group and add admin for enabling future deletion of this contact + groups
+      this.#addAdminInMem(updatedContactGroup.value, linkedName);
+      await this.#updateGroup(updatedContactGroup.id, updatedContactGroup.value, linkedIdkeys);
+    }
+  
+    await this.#linkGroups(Higher.CONTACTS, contactName, linkedIdkeys);
+  }
+  
+  /**
+   * Shares own contact info and requests the contact info of contactIdkey.
+   * TODO implement private contact discovery and return contact name.
+   *
+   * @param {string} contactIdkey hex-formatted public key
+   */
+  async addContact(contactIdkey: string) {
+    // only add contact if not self
+    let linkedName = this.getLinkedName();
+    if (!this.#isMember(contactIdkey, [linkedName])) {
+      // piggyback own contact info when requesting others contact info
+      await this.#requestContact(linkedName, this.#getAllSubgroups([linkedName]), [contactIdkey]);
+    } else {
+      this.#printBadContactError();
+    }
+  }
+
+  /**
+   * Remove contact.
+   *
+   * @param {string} name contact name
+   */
+  async removeContact(name: string) {
+    await this.#deleteGroup(name, this.#resolveIDs([Higher.LINKED]));
+  }
+  
+  /**
+   * Get all contacts.
+   *
+   * @returns {string[]}
+   */
+  getContacts(): string[] {
+    return this.#getChildren(Higher.CONTACTS);
+  }
+  
+  /**
+   * Get pending contacts.
+   * TODO implement pending list.
+   *
+   * @returns {string[]}
+   */
+  getPendingContacts(): string[] {
+    return [];
+  }
+
+  /*
+   ****************
+   * Data methods *
+   ****************
+   */
+  
+  /**
+   * Get storage key for item given prefix and id.
+   *
+   * @param {string} prefix key prefix (GROUPS or app-specific)
+   * @param {string} id auto-incremented id
+   * @returns {string}
+   *
+   * @private
+   */
+  #getDataKey(prefix: string, id: string): string {
```

Comment:
personally think we should just take a "key" as a parameter rather than prefix and id. This should be about mapping keys as they are seen by a developer to as they are seen by our lib.  So it'll just look something like this:

`#getDataKey(key: string): string {
    return  data_prefix + "/" + key + "/";
}`

I think on the developer end since they're still interacting with the lib like it is a key value store, it is not that much of a stretch that they could concat prefix + id by themselves, and then this would generalize more, like if a developer wanted no app-prefixes and human-readable Ids for objects, like in the lightbulb case